File: sdk-py/src/boxty/sandbox.py
from __future__ import annotations
from typing import Any, List, Optional, Callable
from contextlib import contextmanager

class Sandbox:
    def __init__(self, client, data: dict):
        self._client = client
        self._data = data

    def __getattr__(self, name: str) -> Any:
        return self._data.get(name)

    @property
    def id(self) -> str:
        return self._data.get("id")

    @property
    def status(self) -> str:
        return self._data.get("status")

    @property
    def url(self) -> Optional[str]:
        return self._data.get("url")

    @property
    def ws_url(self) -> Optional[str]:
        return self._data.get("ws_url")

    def exec(self, command: str, timeout: int = 60) -> Any:
        return self._client._request("POST", f"/api/sandboxes/{self.id}/exec", json={"command": command, "timeout": timeout})

    @contextmanager
    def forward(self, port: int):
        data = self._client._request("POST", f"/api/sandboxes/{self.id}/forward", json={"port": port})
        yield data.get("url")

    def checkpoint(self, name: str) -> Any:
        return self._client._request("POST", f"/api/sandboxes/{self.id}/snapshot", json={"name": name})

    def stop(self) -> None:
        self._client._request("DELETE", f"/api/sandboxes/{self.id}")

    def metrics(self) -> Any:
        return self._client.get_sandbox_metrics(self.id)

    def attach_secrets(self, secret_names: List[str]) -> Any:
        return self._client.attach_secrets(self.id, secret_names)

    def on(self, event: str, callback: Callable) -> None:
        # Stub for WebSocket event handling
        pass

File: sdk-js/src/sandbox.ts
import { EventEmitter } from 'events';
import { Client } from './client.js';
import { ExecResult, SandboxMetrics } from './types.js';

export class Sandbox extends EventEmitter {
  private client: Client;
  private data: any;

  constructor(client: Client, data: any) {
    super();
    this.client = client;
    this.data = data;
  }

  get id(): string { return this.data.id; }
  get status(): string { return this.data.status; }
  get url(): string { return this.data.url; }
  get startedAt(): string | undefined { return this.data.started_at; }
  get finishedAt(): string | undefined { return this.data.finished_at; }
  get bootDurationMs(): number | undefined { return this.data.boot_duration_ms; }

  async exec(command: string, timeout?: number): Promise<ExecResult> {
    return this.client['request']('POST', `/api/sandboxes/${this.id}/exec`, { command, timeout });
  }

  async forward(port: number): Promise<string> {
    const data = await this.client['request']('POST', `/api/sandboxes/${this.id}/forward`, { port });
    return data.url;
  }

  async checkpoint(name: string): Promise<any> {
    return this.client['request']('POST', `/api/sandboxes/${this.id}/snapshot`, { name });
  }

  async stop(): Promise<void> {
    await this.client.deleteSandbox(this.id);
  }

  async metrics(): Promise<SandboxMetrics> {
    return this.client.getSandboxMetrics(this.id);
  }

  async attachSecrets(secretNames: string[]): Promise<void> {
    await this.client.attachSecrets(this.id, secretNames);
  }

  static async restore(snapshotKey: string): Promise<Sandbox> {
    throw new Error('restore not implemented');
  }
}
